src/libraries/System.Net.Primitives/src/System/Net/Cookie.cs

```diff
@@ -307,6 +307,34 @@ private static bool IsDomainEqualToHost(string domain, string host)
                    (string.Compare(host, 0, domain, 1, host.Length, StringComparison.OrdinalIgnoreCase) == 0);
         }
 
+        internal static int CompareDomain(string? left, string? right)
+        {
+            if (left is null)
+            {
+                return right is null ? 0 : -1;
+            }
+
+            if (right is null)
+            {
+                return 1;
+            }
+
+            int indexLeft = GetComparisonStartIndex(left);
+            int indexRight = GetComparisonStartIndex(right);
+
+            int result;
+
+            if ((result = (left.Length - indexLeft).CompareTo(right.Length - indexRight)) != 0)
+            {
+                return result;
+            }
```

Comment:
Shouldn't this delegate to `string.Compare` right after calculating offsets?

String comparison will first look at the contents, with length only used in case the substrings are equal.

Reply:
Did I get it right that you suggest to remove this check? My idea was to prevent case when right domain is longer than left one and their substring of length `left.Length` are equal i.e. example and example.com

Reply:
> Did I get it right that you suggest to remove this check?

Yes.

> My idea was to prevent case when right domain is longer than left one and their substring of length left.Length are equal

`string.Compare` takes care of this case. It will see that `example` and `example.com` are equal on the shortest substring and then fallback to comparing original lengths (7 vs 11).


Doing the length check upfront would change the sorting order:
`aa, bb, c, d` should sort into `aa, bb, c, d`, but if you check the length first, it would be `c, d, aa, bb`.

Reply:
Hmmm, initial `string.Compare(string, string, StringComparison)` do take care about length, but the current one `string.Compare("example", 0, "example.com", 0, "example".Length, StringComparison.OrdinalIgnoreCase);` returnes 0 in described case.
But I see your point about sorting order. I just find out that `CookieComparer.Compare` is used only for equality checks - so it seems that we can remove logic regarding sorting order.

Reply:
In that case, using `Span`s may be easier:

```c#
public int Compare(ReadOnlySpan<char> left, ReadOnlySpan<char> right)
{
    if (left.StartsWith('.')) left = left.Slice(1);
    if (right.StartsWith('.')) right = right.Slice(1);

    return left.CompareTo(right, StringComparison.OrdinalIgnoreCase);
}
```

Reply:
Replaced